internal/storage/v2/v1adapter/factory.go

```diff
@@ -16,23 +15,61 @@ type Factory struct {
 	ss storage_v1.Factory
 }
 
-func NewFactory(ss storage_v1.Factory) *Factory {
-	return &Factory{
+func NewFactory(ss storage_v1.Factory) tracestore.Factory {
+	factory := &Factory{
 		ss: ss,
 	}
-}
 
-// Initialize implements tracestore.Factory.
-func (*Factory) Initialize(_ context.Context) error {
-	panic("not implemented")
-}
+	var (
+		purger, isPurger   = ss.(storage_v1.Purger)
+		sampler, isSampler = ss.(storage_v1.SamplingStoreFactory)
+		closer, isCloser   = ss.(io.Closer)
+	)
 
-// Close implements tracestore.Factory.
-func (f *Factory) Close(_ context.Context) error {
-	if closer, ok := f.ss.(io.Closer); ok {
-		return closer.Close()
+	switch {
+	case isPurger && isSampler && isCloser:
```

Comment:
I think Closer can be added unconditionally at the struct level, and only implemented conditionally in the actual Close function

Reply:
it would lead to fewer permutations here, and it's not sensitive to the same runtime cast behavior which is more important for the actual business interfaces.